python/cudf/cudf/pylibcudf_tests/test_copying.py

```diff
@@ -0,0 +1,579 @@
+    with pytest.raises(ValueError):
+        plc.copying.gather(
+            target_table,
+            gather_map,
+            plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+        )
+
+
+def test_scatter_table(source_table, input_column, target_table):
+    result = plc.copying.scatter(
+        source_table,
+        input_column,
+        target_table,
+    )
+    pa_source_table = table_to_arrow(source_table)
+    pa_input_column = column_to_arrow(input_column)
+    pa_target_table = table_to_arrow(target_table)
+
+    # TODO: Is there a cleaner vectorized way to do this in pyarrow?
+    output_rows = []
+    for i in range(target_table.num_rows()):
+        if (source_row := pa.compute.index_in(i, pa_input_column)).is_valid:
+            output_rows.append(pa_source_table.take([source_row]))
+        else:
+            output_rows.append(pa_target_table.take([i]))
+
+    expected = pa.concat_tables(output_rows)
+    assert_table_eq(result, expected)
+
+
+def test_scatter_table_num_col_mismatch(
+    source_table, input_column, target_table
+):
+    # Number of columns in source and target must match.
+    with pytest.raises(ValueError):
+        plc.copying.scatter(
+            plc.Table(source_table.columns()[:2]),
+            input_column,
+            target_table,
+        )
+
+
+def test_scatter_table_num_row_mismatch(source_table, target_table):
+    # Number of rows in source and scatter map must match.
+    with pytest.raises(ValueError):
+        plc.copying.scatter(
+            source_table,
+            column_from_arrow(pa.array(range(source_table.num_rows() * 2))),
+            target_table,
+        )
+
+
+def test_scatter_table_map_has_nulls(source_table, target_table):
+    with pytest.raises(ValueError):
+        plc.copying.scatter(
+            source_table,
+            column_from_arrow(pa.array([None] * source_table.num_rows())),
+            target_table,
+        )
+
+
+def test_scatter_table_type_mismatch(source_table, input_column, target_table):
+    with pytest.raises(TypeError):
+        pa_array = pa.array([True] * source_table.num_rows())
+        ncol = source_table.num_columns()
+        pa_table = pa.table([pa_array] * ncol, [""] * ncol)
+        plc.copying.scatter(
+            plc.Table.from_arrow(pa_table),
+            input_column,
+            target_table,
+        )
+
+
+def test_scatter_scalars(source_scalar, input_column, target_table):
+    result = plc.copying.scatter(
+        [source_scalar] * target_table.num_columns(),
+        input_column,
+        target_table,
+    )
+    # TODO: There's no reason to require column metadata in this API. We can construct a
+    # placeholder internally and discard it after the conversion if necessary.
+    host_scalar = scalar_to_arrow(source_scalar).as_py()
+    pa_input_column = column_to_arrow(input_column)
+    pa_target_table = table_to_arrow(target_table)
+
+    # TODO: Is there a cleaner vectorized way to do this in pyarrow?
```

Comment:
As it gets more complex to programmatically generate the result, I lean more towards hand coding some output data - it might make it clearer from a glance exactly what we expect to produce and save some of the effort later on when it comes to things like groupby and joins.

Reply:
That is true. The tradeoff is that with the current approach we verify against some gold standard (pyarrow in this case) rather than our own expectation, and we get something that automatically adapts to different types of data etc, making test parametrization easier. There's definitely a balance to be struck here. Let me see if I can find good ways to simplify some of the more complex tests, otherwise we can certainly reconsider hardcoding values in some places.

Reply:
We discussed this offline. Now that pyarrow actually implements a lot of what we need for testing, we're going to move forward with testing against a pyarrow implementation where possible rather than hardcoding values.